pico/configdriven/services/src/main/java/io/helidon/pico/configdriven/services/AbstractConfiguredServiceProvider.java

```diff
@@ -0,0 +1,832 @@
+        rootProvider.set(null);
+        initialized.set(false);
+        initializationCallingContext.set(null);
+        return true;
+    }
+
+    void assertIsInitializing() {
+        if (initialized.get()) {
+            CallingContext callingContext = initializationCallingContext.get();
+            String desc = description() + " was previously initialized";
+            String msg = (callingContext == null) ? toErrorMessage(desc) : toErrorMessage(callingContext, desc);
+            throw new PicoServiceProviderException(msg, this);
+        }
+    }
+
+    void assertIsInitialized() {
+        if (!initialized.get()) {
+            throw new PicoServiceProviderException(description() + " was expected to be initialized", this);
+        }
+    }
+
+    @Override
+    protected void doPreDestroying(LogEntryAndResult logEntryAndResult) {
+        if (isRootProvider()) {
+            managedConfiguredServicesMap.values().stream()
+                    .filter(Optional::isPresent)
+                    .map(Optional::get)
+                    .forEach(csp -> {
+                        LogEntryAndResult cspLogEntryAndResult = csp.createLogEntryAndResult(Phase.DESTROYED);
+                        csp.doPreDestroying(cspLogEntryAndResult);
+                    });
+        }
+        super.doPreDestroying(logEntryAndResult);
+    }
+
+    @Override
+    protected void doDestroying(LogEntryAndResult logEntryAndResult) {
+        super.doDestroying(logEntryAndResult);
+    }
+
+    @Override
+    protected void onFinalShutdown() {
+        if (isRootProvider()) {
+            managedConfiguredServicesMap.values().stream()
+                    .filter(Optional::isPresent)
+                    .map(Optional::get)
+                    .filter(csp -> csp.currentActivationPhase().eligibleForDeactivation())
+                    .forEach(AbstractConfiguredServiceProvider::onFinalShutdown);
+        }
+
+        this.initialized.set(false);
+        this.managedConfiguredServicesMap.clear();
+        this.configBeanMap.clear();
+
+        super.onFinalShutdown();
+    }
+
+    /**
+     * Transition into an initialized state.
+     */
+    void assertInitialized(boolean initialized) {
+        assertIsInitializing();
+        assert (!drivesActivation()
+                        || isAlreadyAtTargetPhase(PicoServices.terminalActivationPhase())
+                        || managedConfiguredServicesMap.isEmpty());
+        this.initialized.set(initialized);
+    }
+
+    /**
+     * Maybe transition into being a root provider if we are the first to claim it. Otherwise, we are a slave being managed.
+     *
+     * @param isRootProvider    true if an asserting is being made to claim root or claim managed slave
+     * @param expectSet         true if this is a strong assertion, and if not claimed an exception will be thrown
+     */
+    // special note: this is referred to in code generated code!
+    protected void assertIsRootProvider(boolean isRootProvider,
```

Comment:
Wrong order of methods within a class.

Reply:
still static method is last.
also inner classes should be the very last

Reply:
pushed